eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java

```diff
@@ -0,0 +1,399 @@
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(key, values);
+        }
+    }
+
+    @Override
+    public <KK extends K, VV extends V> boolean putAll(Multimap<KK, VV> multimap)
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(multimap);
+        }
+    }
+
+    @Override
+    public void clear()
+    {
+        synchronized (this.lock)
+        {
+            this.getDelegate().clear();
+        }
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.isEmpty();
+        }
+    }
+
+    @Override
+    public boolean notEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.notEmpty();
+        }
+    }
+
+    @Override
+    public void forEachValue(Procedure<? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKey(Procedure<? super K> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKey(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyValue(Procedure2<? super K, ? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyMultiValues(Procedure2<? super K, ? super Iterable<V>> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyMultiValues(procedure);
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.size();
+        }
+    }
+
+    @Override
+    public int sizeDistinct()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.sizeDistinct();
+        }
+    }
+
+    @Override
+    public boolean containsKey(Object key)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKey(key);
+        }
+    }
+
+    @Override
+    public boolean containsValue(Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsValue(value);
+        }
+    }
+
+    @Override
+    public boolean containsKeyAndValue(Object key, Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKeyAndValue(key, value);
+        }
+    }
+
+    @Override
+    public RichIterable<K> keysView()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keysView();
+        }
+    }
+
+    @Override
+    public SetIterable<K> keySet()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keySet();
+        }
```

Comment:
Can you cast?
On Sun, Aug 14, 2016 at 8:38 AM Hiroshi Ito <notifications@github.com>
wrote:

> In
> eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java
> <https://github.com/eclipse/eclipse-collections/pull/93#discussion_r74701062>
> :
>
> > +    @Override
> > +    public RichIterable<K> keysView()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keysView();
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public SetIterable<K> keySet()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keySet();
> > +        }
>
> The tricky part here is that the return type of Multimap.keySet() is
> supposed to be SetIterable (refer to issue #45
> <https://github.com/eclipse/eclipse-collections/issues/45>). There is no
> unmodifiable wrapper or synchronized wrapper that takes SetIterable as an
> input parameter, so we'd need to implement ones in order to do so. I guess
> we can address this in a separate issue before 8.0 official release. It's
> just that I think it's too much for @g-votte <https://github.com/g-votte>
> to take up in this same issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/pull/93/files/ace474519c1dbe03d2783ec5640c911f58ada19d..7de3f8ba027df68f6a4eb3621877b9007dbe0c11#r74701062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO6Iq_Ykjo29HoKDUENVdmKC7BzkRg-ks5qfwxFgaJpZM4ILYAA>
> .
>